test/Ocelot.UnitTests/DownstreamRouteFinder/UrlMatcher/UrlPathPlaceholderNameAndValueFinderTests.cs

```diff
@@ -349,6 +349,20 @@ public void can_match_down_stream_url_with_downstream_template_with_place_holder
                  .BDDfy();
         }
 
+        [Fact]
+        public void check_for_placeholder_at_end_of_template() 
+        {
+            var expectedTemplates = new List<PlaceholderNameAndValue>
+            {
+                new("{testId}", string.Empty),
+            };
+            this.Given(x => x.GivenIHaveAUpstreamPath("/upstream/test/"))
+                .And(x => x.GivenIHaveAnUpstreamUrlTemplate("/upstream/test/{testId}"))
+                .When(x => x.WhenIFindTheUrlVariableNamesAndValues())
+                .And(x => x.ThenTheTemplatesVariablesAre(expectedTemplates))
+                .BDDfy();
+        }
```

Comment:
I know other unit tests can cover our user scenarios, and acceptance tests can cover too, **but...**
For bugs it is better to write one `Theory` test with name `should_fix_748_issue` with all templates from descriptions, author's and contributor's ones.

### User scenario 1

```
"DownstreamPathTemplate": "/api/invoices/{url}"
"UpstreamPathTemplate": "/{url}"
```

It maps URLs this way:

```
/ → /api/invoices/
/123 → /api/invoices/123
```

### User scenario 2

```
"DownstreamPathTemplate": "/api/invoices/{url}"
"UpstreamPathTemplate": "/test/{url}"
```

It works correctly when `{url}` is specified: `/test/123 → /api/invoices/123`

But it fails, when `{url}` is empty. I would expect upstream URL `/test/` to reroute to downstream URL `/api/invoices/`.

### User scenario 3

```
"DownstreamPathTemplate": "/api/invoices{url}"
"UpstreamPathTemplate": "/test{url}"
```

It correctly maps non epmty `{url}` which means, it works with the slash at the end, because the slash is the value of the `{url}`. But still not without it:

```
/test/123 → /api/invoices/123
/test/ → /api/invoices/
/test → Failed to match ReRoute configuration for upstream path: /test
```

To make work everything what I want, I need three reroutes (or at least two, if the first one will be without slashes):

```
{
    "DownstreamPathTemplate": "/api/invoices/{url}"
    "UpstreamPathTemplate": "/test/{url}"
},
{
    "DownstreamPathTemplate": "/api/invoices/"
    "UpstreamPathTemplate": "/test/"
},
{
    "DownstreamPathTemplate": "/api/invoices"
    "UpstreamPathTemplate": "/test"
}
```
See [original description](https://github.com/ThreeMammals/Ocelot/issues/748#issue-399043741).

Reply:
For placeholders in the middle, please write one extra `Theory` with unit tests 👉 
> [satano](https://github.com/satano) commented [on Jan 15, 2019](https://github.com/ThreeMammals/Ocelot/issues/748#issuecomment-454305176)

Reply:
@raman-m all the unit tests exist as part of the routing tests
https://github.com/ThreeMammals/Ocelot/pull/1911/files#diff-99c1f697736fe2348d794d9911adf9dde5428205b123b99511fab187260ffd60L810